### ISSUE TITLE
What should be the delimiter of identifiers in file names?

### ISSUE THREAD

**Issue by jeanphilippegg** [protesilaos/denote#332]
Ahead of the change that makes it possible to have file name components in another order, I would like to have opinions on what should be the delimiter of identifiers when it does not appear as the first component of the file name.

In my fork of the code, I have reused the signature delimiter (==), but I am not sure that reusing an existing delimiter is a good idea. I am thinking of using ++ instead. It is already sluggified away by default. Any objections? Ideas?

For example, we could have `My-note==sig1__emacs_philosophy++20240303T030303.org`.

**Comment by protesilaos** [comment]
> From: Jean-Philippe Gagné Guay ***@***.***>
> Date: Fri, 26 Apr 2024 17:37:19 -0700
>
> Ahead of the change that makes it possible to have file name
> components in another order, I would like to have opinions on what
> should be the delimiter of identifiers when it does not appear as the
> first component of the file name.
>
> In my fork of the code, I have reused the signature delimiter (==),
> but I am not sure that reusing an existing delimiter is a good idea. I
> am thinking of using ++ instead. It is already sluggified away by
> default. Any objections? Ideas?
>
> For example, we could have `My-note==sig1__emacs_philosophy++20240303T030303.org`.

Keep the plus for the signature, otherwise it will be
backward-incompatible.

The plus sign looks fine. Though it can be used in regexp, which makes
it a bit less convenient. Same with the asterisk.

Other options are the hash (#), semicolon (;), at (@), dollar ($), caret
(^). I omit the ampersand because of its meaning in the Unix shell.

Of those I like the # and the @, with the $ as a close third.

I know Emacs uses the # for its temporary recovery files, though those
appear at the start and end of the name. I don't think we have a problem
there.

I am not aware of other incompatibilities, though we need to be careful
because this is a choice we do not want to undo months later when people
have written their files already (plus links, Org dynamic blocks, ...).

-- 
Protesilaos Stavrou
https://protesilaos.com

**Comment by jeanphilippegg** [comment]
> Keep the plus for the signature, otherwise it will be backward-incompatible.

What do you mean by that? Did you mean to keep the equal sign for the signature? If so, my initial reasoning was that identifiers have a rigid format so it would not conflict with the == of the signature. It would be unlikely that someone wants to create a signature that has the format of an identifier. However, I agree that we should be using a new delimiter instead.

> this is a choice we do not want to undo months later when people have written their files already (plus links, Org dynamic blocks, ...).

Yes, we need to be careful before making this change. This change will only affect the way the identifier appears in the file name (if it does not appear as the first component). Links, Org dynamic blocks, etc. should be unaffected.

I also like # and @. Do you have a preference between +, # and @?

**Comment by newmedicine** [comment]
Somehow using # feels messy because of the temp files issue

**Comment by protesilaos** [comment]
> What do you mean by that? Did you mean to keep the equal sign for the signature? 

Yes, sorry!

> Do you have a preference between +, # and @?

No preference. Though the temp files, as also noted by @newmedicine, as a good argument against it. So I would then check if there is any good argument against the @.

**Comment by maikol-solis** [comment]
Having a file scheme like 

```
SIGNATURE--TITLE__KEYWORDS@@ID
```

will be very useful, specially for people like me, that everything is ordered using the Luhmann's (Zettelkasten) system. 

I can test and report any bug once it lands to master or another feature branch. 

Best.

**Comment by mentalisttraceur** [comment]
I would suggest `++` over `@@` because a delimiter which is smaller/sparcer than the text it delimits usually makes it easier to spot the boundary, even at a glance, or from a distance, or without focusing eyes on it. Compare:

![Screenshot_20240516-165605](https://github.com/protesilaos/denote/assets/6531708/217b750b-e503-49f1-8310-5c7ca981c6a1)

Admittedly, this is somewhat font-relative, and `@` can stand out in its own way, by being bigger/denser than most other characters.

But I would give my opinion reduced weight because I'm probably always going to keep the identifier as the first component, so I'm not going to have to live with the choice as much as others will.

**Comment by jeanphilippegg** [comment]
Given the idea I shared in #278 about the identifier that could become
separate from the date, maybe we will actually need 2 new delimiters instead
of 1.

I also like `++`. One objection against it is that it is special in Emacs
regexps and it would need to be escaped. However, I still like it better
visually. Moreover, Emacs' regexps syntax is already a little hard to parse.

If we are to go with what I described in #278, we could use `@@` for the id and
`++` for the date, or the opposite.

**Comment by protesilaos** [comment]
> From: mentalisttraceur ***@***.***>
> Date: Thu, 16 May 2024 14:24:06 -0700
>
> I would suggest `++` over `@@` because a delimiter which is
> smaller/sparcer than the text it delimits usually makes it easier to
> spot the boundary, even at a glance, or from a distance, or without
> focusing eyes on it. Compare:
>
> ![Screenshot_20240516-165605](https://github.com/protesilaos/denote/assets/6531708/217b750b-e503-49f1-8310-5c7ca981c6a1)
>
> Admittedly, this is somewhat font-relative, and `@` can stand out in
> its own way, by being bigger/denser than most other characters.

Here it does look better, though it is indeed dependent on the font. My
concern with the choice of character is to avoid special regexp
characters. This includes the asterisk, caret, dollar, and question
mark. Using those is technically fine though it adds some extra friction
while performing a search.

> But I would give my opinion reduced weight because I'm probably always
> going to keep the identifier as the first component, so I'm not going
> to have to live with the choice as much as others will.

I won't move the identifiers either, though I want this to be as easy to
use as the other delimiters we have. Simply because the user will not
have to think twice about when to escape a character or not. Regexp
notation is already complex and difficult for many people as it is.

-- 
Protesilaos Stavrou
https://protesilaos.com

**Comment by protesilaos** [comment]
> From: Jean-Philippe Gagné Guay ***@***.***>
> Date: Thu, 16 May 2024 20:05:00 -0700
>
> Given the idea I shared in #278 about the identifier that could become
> separate from the date, maybe we will actually need 2 new delimiters instead
> of 1.

My concern is this will make things more complex. A separate date file
name component comes with own extra set of functions for retrieving that
data. Then the renaming will have to account for it. File names with
optional fields will have more permutations. The fontification will be
different. And we will have to decide what to do with cases such as
linking to a file with its date, whether the date should change freely
or not, and so on. Do we have a good use-case for going down this path?

Right now, people who really need to store the date in a separate
component can use the signature. We added the signature as a catch-all
field. Yes, the motivation was to support the use-case of Luhmann-style
identifier notation, though we do nothing to enforce such a scheme: the
signature takes any content (modulo the sluggification).

About the identifier having a different scheme while retaining its
current dual function as date+time and ID, I think this is easier to do.
It will basically keep things as they are but add more flexibility
inside the identifier component. Though we need to think carefully how
best to expose this as an option, so that we can still have (i)
identifiers that are easy to predict and (ii) backward-compatibility.

> I also like `++`. One objection against it is that it is special in Emacs
> regexps and it would need to be escaped. However, I still like it better
> visually. Moreover, Emacs' regexps syntax is already a little hard to parse.

Yes, this is the gist of it. I think it is better to avoid all special
regexp characters, including those that could look nicer like the
dollar.

-- 
Protesilaos Stavrou
https://protesilaos.com

**Comment by mentalisttraceur** [comment]
> it adds some extra friction while performing a search.

Oh! You're right, this is a *user-facing* inconvenience for anyone using regex search. Sorry, only thought about having to escape it in code, but this would affect i-search, consult/ivy/helm (if configured to only use regex match), and so on.

I now agree that `@@` is the better choice.
I retract my suggestion of `++`.

**Comment by maikol-solis** [comment]
Just a weird suggestion, but why not use something like `=-`or `-=` for the ID instead? 

Is it mandatory to use the same character twice? 

For example, some variations of the same file. 
```
my-title==QUX=-20240517T102209_keyword.md
my-title==QUX-=20240517T102209_keyword.md

QUX--my-title=-20240517T102209_keyword.md
QUX--my-title-=20240517T102209_keyword.md
```

**Comment by mentalisttraceur** [comment]
Oh, I forgot to say, `@` has a nice mnemonic/semantic connection to linking and identifiers: we already use `@foo` in day-to-day life to refer/link to `foo`.

**Comment by protesilaos** [comment]
> From: Maikol Solís ***@***.***>
> Date: Fri, 17 May 2024 09:28:18 -0700
>
> Just a weird suggestion, but why not use something like `=-`or `-=` for the ID instead? 
>
> Is it mandatory to use the same character twice? 

Since we do it for the other components, it is easier to keep things
consistent.

For the file components apart from the identifier, the double sign is a
way to match a string at the beginning of the field, like "--word", as
opposed to anywhere in the field with "-word". The dash here limits the
search to the title component of the file name (assuming default
settings). This is done without knowledge of regular expressions.

-- 
Protesilaos Stavrou
https://protesilaos.com

**Comment by protesilaos** [comment]
> From: mentalisttraceur ***@***.***>
> Date: Fri, 17 May 2024 12:04:44 -0700
>
> Oh, I forgot to say, `@` has a nice mnemonic/semantic connection to linking and identifiers: we already use ***@***.***` in day-to-day life to refer/link to `foo`.

Indeed!

-- 
Protesilaos Stavrou
https://protesilaos.com
